### PR TITLE
Rewrite Cardano's `Transaction` and `Address` implementations

### PR DESCRIPTION
**Pull request by iokasimov**
Changes needed in [Payment module in P2P](https://github.com/zkFold/p2p-onramp/blob/main/src/ZkFold/P2P/Payment.hs)
Related task: https://github.com/zkFold/zkfold-base/issues/10

### CODE REVIEW

File: src/ZkFold/Symbolic/Cardano/Types.hs [hunk 1]
```diff
@@ -1,13 +1,101 @@
...
+
+deriving instance
+    ( KnownNat n
+    , Arithmetizable i a
+    , Arithmetizable i (UInt 64 a)
+    , Arithmetizable i (ByteString 8 a)
+    , Arithmetizable i (ByteString 15 a)
+    ) => Arithmetizable i (Value n a)
+
+newtype Input datum a = Input (OutputRef a, (a, datum a))
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Cardano/Types.hs:56]: Inputs should be pairs of (OutputRef, Output).

File: src/ZkFold/Symbolic/Cardano/Types.hs [hunk 2]
```diff
@@ -1,13 +1,101 @@
...
+txInputs :: Transaction inputs rinputs outputs tokens datum a -> Vector inputs (Input datum a)
+txInputs (Transaction (_, (is, _))) = is
+
+txOutputs :: Transaction inputs rinputs outputs tokens datum a -> Vector outputs (Output tokens a)
+txOutputs (Transaction (_, (_, (os, _)))) = os
+
+newtype TxId a = TxId a
+    deriving (Eq, Arithmetizable i)
+
+newtype Value n a = Value (Vector n (ByteString 8 a, (ByteString 15 a, UInt 64 a)))
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Cardano/Types.hs:45]: 8 and 15 are the number of bits in `CurrencySymbol` and `TokenName`. respectively. It should be 224 bit for the `CurrencySymbol`, and either 256 or 512 bit for the `TokenName` (please check).

File: src/ZkFold/Symbolic/Cardano/Types.hs [hunk 3]
```diff
@@ -1,13 +1,101 @@
...
+    ) => Arithmetizable i (Output tokens a)
+
+newtype OutputRef a = OutputRef (TxId a, UInt 32 a)
+    deriving Eq
+
+deriving instance
+    (Arithmetizable i a, Arithmetizable i (UInt 32 a))
+    => Arithmetizable i (OutputRef a)
+
+newtype Address a = Address (ByteString 3 a, (ByteString 28 a, ByteString 28 a))
```
Comment (vlasin) [inline src/ZkFold/Symbolic/Cardano/Types.hs:89]: In `ByteString n a`, `n` is the number of bits! Both credentials are 224-bit.

Comment (iokasimov) [reply src/ZkFold/Symbolic/Cardano/Types.hs]: @vlasin Sorry, I mistakenly assumed without asking that it's a byte string length.

File: src/ZkFold/Symbolic/Cardano/Types.hs (hunk 2 above)
Comment (iokasimov) [reply src/ZkFold/Symbolic/Cardano/Types.hs]: @vlasin `TokenName` should be no longer than 32 bytes so it's `256`.

**Review by vlasin — APPROVED** [review]

**Pull request merged**